File: Algorithm/QuickSort.m
% 快速排序，p,r为待排序数组下标，原址
function Array=QuickSort(Array,p,r)
if p<r
    [Array,q]=sub_ranPartition(Array,p,r);
    Array=QuickSort(Array,p,q-1);
    Array=QuickSort(Array,q+1,r);
end
end

function [A,q]=sub_partition(A,p,r)
x=A(r);
i=p-1;
for j=p:r-1
    if A(j)<=x
        i=i+1;
        temp=A(i);
        A(i)=A(j);
        A(j)=temp;
    end
end
temp=A(i+1);
A(i+1)=A(r);
A(r)=temp;
q=i+1;
end

function [A,q]=sub_ranPartition(A,p,r)
i=round(p+(r-p)*rand());
temp=A(i);
A(i)=A(r);
A(r)=temp;
[A,q]=sub_partition(A,p,r);
end

File: Algorithm/Sort & Order Statistic/C/QuickSort.c
int sub_partition(int* A, int p, int r)
{
	int x = *(A + r - 1);
	int j, i = p - 2;
	int temp;
	for (j = p - 1; j < r - 1; j++)
		if (*(A + j) <= x)
		{
			i++;
			temp = *(A + i);
			*(A + i) = *(A + j);
			*(A + j) = temp;
		}
	temp = *(A + i + 1);
	*(A + i + 1) = *(A + r -1);
	*(A + r - 1) = temp;
	int q = i + 1;
	return q;
}

int sub_ranPartition(int* A, int p, int r)
{
	srand((unsigned)time(NULL));    //产生伪随机数序列的种子，应放在主程序中
	int temp;
	int i = rand() % (r - p + 1) + p;
	printf("%d\n", i);
	temp = *(A + i - 1);
	*(A + i - 1) = *(A + r - 1);
	*(A + r - 1) = temp;
	int q = sub_partition(A, p, r);
	return q;
}

void QuickSort(int* A, int p, int r)
{
	if (p < r)
	{
		int q = sub_ranPartition(A, p, r);
		QuickSort(A, p, q - 1);
		QuickSort(A, p + 1, r);
	}
}